(8.0) PXC-3576 PXC create a role mapping an upper case database with lower_…

aybek opened the pull request:
…case_table_names=1

Temporary workaround until the issue is fixed on PXB

aybek commented:
https://pxc.cd.percona.com/view/8.0%20parallel%20MTR/job/pxc-8.0-pipeline-parallel-mtr/3169/

kamil-holubicki commented on scripts/mysql_system_tables_fix.sql:
```diff
@@ -1415,8 +1415,30 @@ INSERT IGNORE INTO mysql.user VALUES ('localhost','mysql.pxc.sst.role','N','N','
+
+    IF lctn = 1 THEN
+        -- Case-insensitive system: use lowercase schema/table
+    INSERT IGNORE INTO mysql.tables_priv VALUES ('localhost', 'percona_schema', 'mysql.pxc.sst.role', 'xtrabackup_history', 'root\@localhost', CURRENT_TIMESTAMP, 'Alter,Select,Insert,Create', '');
```
I think there is something wrong with indentation here and in following conditional branches.

kamil-holubicki commented on scripts/mysql_system_tables_fix.sql:
```diff
@@ -1415,8 +1415,30 @@ INSERT IGNORE INTO mysql.user VALUES ('localhost','mysql.pxc.sst.role','N','N','
-INSERT IGNORE INTO mysql.tables_priv VALUES ('localhost', 'PERCONA_SCHEMA', 'mysql.pxc.sst.role', 'xtrabackup_history', 'root\@localhost', CURRENT_TIMESTAMP, 'Alter,Select,Insert,Create', '');
-INSERT IGNORE INTO mysql.db VALUES ('localhost', 'PERCONA_SCHEMA', 'mysql.pxc.sst.role','N','N','N','N','Y','N','N','N','N','N','N','N','N','N','N','N','N','N','N');
+DELIMITER $$
+DROP PROCEDURE IF EXISTS conditional_grants $$
```
Can we use it with standard delimiter and redefine delimiter after it (swap this and above lines)?

kamil-holubicki commented on scripts/mysql_system_tables_fix.sql:
```diff
@@ -1415,8 +1415,30 @@ INSERT IGNORE INTO mysql.user VALUES ('localhost','mysql.pxc.sst.role','N','N','
-INSERT IGNORE INTO mysql.db VALUES ('localhost', 'PERCONA_SCHEMA', 'mysql.pxc.sst.role','N','N','N','N','Y','N','N','N','N','N','N','N','N','N','N','N','N','N','N');
+DELIMITER $$
+DROP PROCEDURE IF EXISTS conditional_grants $$
+CREATE PROCEDURE conditional_grants()
```
Not a strong opinion. To consider:
In the case of conditional execution of the query accros the file, we use the approach like in line [1081](https://github.com/percona/percona-xtradb-cluster/blob/9549ba275ea8f8fafe09d5521c1b724ff60faaa1/scripts/mysql_system_tables_fix.sql#L1081) which gathers the condition then IFs cmd basing on condition, then creates a prepared stmt, executes, drops.
I don't know if it would be more readable or not, that's why it is not a strong opinion :)

aybek replied:
Left as it is, this way it is more readable to me

kamil-holubicki commented on scripts/mysql_system_users.sql:
```diff
@@ -99,7 +99,37 @@ REVOKE ALL PRIVILEGES, GRANT OPTION FROM 'mysql.pxc.sst.role'@localhost;
 GRANT LOCK TABLES, PROCESS, RELOAD, REPLICATION CLIENT, SUPER, INNODB_REDO_LOG_ARCHIVE ON *.* TO 'mysql.pxc.sst.role'@localhost;
-GRANT ALTER, CREATE, SELECT, INSERT ON PERCONA_SCHEMA.xtrabackup_history TO 'mysql.pxc.sst.role'@localhost;
+
+DELIMITER $$
```
likewise
(all comments from the other file apply here as well)

kamil-holubicki reviewed (changes requested):
Would it be possible to create MTR testcase?

aybek committed:
PXC-3576 PXC create a role mapping an upper case database with lower_case_table_names=1

Problem:
The PXC startup script grants the mysql.pxc.sst.role role access to the PERCONA_SCHEMA database to support PXB’s history feature.
However, when the lower_case_table_names = 1 setting is used in a PXC cluster, it results in duplicate role mapping entries —
one for the `PERCONA_SCHEMA` database and another for `percona_schema` database.
This happens because PXB creates the PERCONA_SCHEMA database using uppercase letters.

Solution:
To address this issue, a fix should be applied on the PXB side by renaming the PERCONA_SCHEMA database to lowercase.

As a temporary workaround, during grants in startup script we check whether lower_case_table_names = 1 is set.

If it is, we grant permissions for mysql.pxc.sst.role on the lowercase percona_schema database.
Otherwise, we use the original uppercase PERCONA_SCHEMA.
